### ISSUE TITLE
VictoriaLogs Bug: Read Timeout with stream_context

### ISSUE THREAD

**Issue by worker24h** [VictoriaMetrics/VictoriaMetrics#7637]
### Describe the bug

The LogQL: `_time:5m MarsSendServiceImpl` , I can get result so quickly by UI , **but** append ` | stream_context before 4`,  the response is error, e.g. `Read Timeout`, as follow:
![image](https://github.com/user-attachments/assets/8754b324-7032-48f5-88e5-bcca6434ac8d)

Other issue: https://github.com/VictoriaMetrics/victorialogs-datasource/issues/117

I think the result should response quicky and not timeout

### To Reproduce

As above

### Version

release 1.0.0

**Comment by jiekun** [comment]
Hi. I'm testing similar query in our victorialogs playground but unlucky cannot reproduce with the current dataset. Test query [here](https://play-vmlogs.victoriametrics.com/select/vmui/?#/?query=*+%7C+stream_context+before+4&g0.range_input=12h&g0.end_input=2024-11-25T04%3A21%3A51&g0.relative_time=last_12_hours&accountID=0&projectID=0).

May I ask if you could provide the dataset you are using? 
- It could be exported via the query API, see this [doc](https://docs.victoriametrics.com/victorialogs/faq/#how-to-export-logs-from-victorialogs).
- It would be great if you could share the actual data files under its data folder. See this [doc](https://docs.victoriametrics.com/victorialogs/#backup-and-restore).

**Comment by worker24h** [comment]
> Hi. I'm testing similar query in our victorialogs playground but unlucky cannot reproduce with the current dataset. Test query [here](https://play-vmlogs.victoriametrics.com/select/vmui/?#/?query=*+%7C+stream_context+before+4&g0.range_input=12h&g0.end_input=2024-11-25T04%3A21%3A51&g0.relative_time=last_12_hours&accountID=0&projectID=0).
> 
> May I ask if you could provide the dataset you are using?
> 
> * It could be exported via the query API, see this [doc](https://docs.victoriametrics.com/victorialogs/faq/#how-to-export-logs-from-victorialogs).
> * It would be great if you could share the actual data files under its data folder. See this [doc](https://docs.victoriametrics.com/victorialogs/#backup-and-restore).


The logs content:
```text
2024-11-25T05:44:48.288Z	error	VictoriaMetrics/app/vlinsert/jsonline/jsonline.go:60	jsonline: cannot read line #4620 in /jsonline request: cannot read json line, since its size exceeds -insert.maxLineSizeBytes=262144
2024-11-25T05:45:02.236Z	warn	VictoriaMetrics/app/vlselect/main.go:169	remoteAddr: "172.25.0.101:54726"; requestURI: /select/logsql/query?end=2024-11-25T05%3A44%3A31.972Z&limit=50&query=%2A+%7C+stream_context+before+4&start=2024-11-25T05%3A39%3A31.972Z; the request couldn't be executed in 30.000 seconds; possible solutions: to increase -search.maxQueryDuration=30s; to pass bigger value to 'timeout' query arg
2024-11-25T05:45:07.667Z	error	VictoriaMetrics/app/vlinsert/jsonline/jsonline.go:60	jsonline: cannot read line #5379 in /jsonline request: cannot read json line, since its size exceeds -insert.maxLineSizeBytes=262144
```
Sorry, I can't offer your data directory, it's too big, more than 20G.

Are you from CHINA? Do you have WeChat？ We can communicate by it

**Comment by worker24h** [comment]
> > Hi. I'm testing similar query in our victorialogs playground but unlucky cannot reproduce with the current dataset. Test query [here](https://play-vmlogs.victoriametrics.com/select/vmui/?#/?query=*+%7C+stream_context+before+4&g0.range_input=12h&g0.end_input=2024-11-25T04%3A21%3A51&g0.relative_time=last_12_hours&accountID=0&projectID=0).
> > May I ask if you could provide the dataset you are using?
> > 
> > * It could be exported via the query API, see this [doc](https://docs.victoriametrics.com/victorialogs/faq/#how-to-export-logs-from-victorialogs).
> > * It would be great if you could share the actual data files under its data folder. See this [doc](https://docs.victoriametrics.com/victorialogs/#backup-and-restore).
> 
> The logs content:
> 
> ```
> 2024-11-25T05:44:48.288Z	error	VictoriaMetrics/app/vlinsert/jsonline/jsonline.go:60	jsonline: cannot read line #4620 in /jsonline request: cannot read json line, since its size exceeds -insert.maxLineSizeBytes=262144
> 2024-11-25T05:45:02.236Z	warn	VictoriaMetrics/app/vlselect/main.go:169	remoteAddr: "172.25.0.101:54726"; requestURI: /select/logsql/query?end=2024-11-25T05%3A44%3A31.972Z&limit=50&query=%2A+%7C+stream_context+before+4&start=2024-11-25T05%3A39%3A31.972Z; the request couldn't be executed in 30.000 seconds; possible solutions: to increase -search.maxQueryDuration=30s; to pass bigger value to 'timeout' query arg
> 2024-11-25T05:45:07.667Z	error	VictoriaMetrics/app/vlinsert/jsonline/jsonline.go:60	jsonline: cannot read line #5379 in /jsonline request: cannot read json line, since its size exceeds -insert.maxLineSizeBytes=262144
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/24838476/389383400-3b29e0d2-c69c-4bfc-923c-3eeb5a0de65f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1MTQyMzUsIm5iZiI6MTczMjUxMzkzNSwicGF0aCI6Ii8yNDgzODQ3Ni8zODkzODM0MDAtM2IyOWUwZDItYzY5Yy00YmZjLTkyM2MtM2VlYjVhMGRlNjVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI1VDA1NTIxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE5ZGQ2Y2Q3MGI2ZTVlNzMzZTE5M2ZjZGQwNjM4ZDIyN2Q4MmE2NmFlYzRiZTQ1ZWNjMmM0NGJhMTk4MDBjMmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SvtXGkMMW5A5K8vzX6pPR3M5ABbfXdOKq5iAzlgvMaA)
> 
> Sorry, I can't offer your data directory, it's too big, more than 20G，800 million pieces of data are generated every day
> 
> Are you from CHINA? Do you have WeChat？ We can communicate by it

@jiekun

**Comment by worker24h** [comment]
![image](https://github.com/user-attachments/assets/0d7f78c4-e60c-4e24-baca-22194742e01b)

I executed the `iostat` of linux command in DOCKER container, as above. I think this case is not hit index ？

**Comment by worker24h** [comment]
Hello, The issue is a bug? Who can fixed it?

**Comment by jiekun** [comment]
@worker24h Hello Sorry for the delay. I was able to reproduce it on our dataset. Working on a fix now.

**Comment by worker24h** [comment]
> @worker24h Hello Sorry for the delay. I was able to reproduce it on our dataset. Working on a fix now.

How to reproduce it? What's the root cause?

**Comment by jiekun** [comment]
As an example, please check this logQL on our public VictoriaLogs playground, which goes timeout just like yours: https://play-vmlogs.victoriametrics.com/select/vmui/?#/?query=*+%7C+stream_context+before+4&g0.range_input=1h&g0.end_input=2024-11-25T03%3A37%3A26&g0.relative_time=none

To confim, may I ask:
1. how many logs match the logQL `without stream_context before 4`? (which is `_time:5m MarsSendServiceImpl` in your case)?
2. how many memory resource did you allocate to this VictoriaLogs instance?

It seems VictoriaLogs is retrieving too many streams before goes to the `stream_context` pipe and cost a lot of time. I'm still investigating.

<details>

![image](https://github.com/user-attachments/assets/9cce87e0-baaf-4a33-ab95-42d72d098bb3)

</details>

**Comment by worker24h** [comment]
Q1: how many logs match the logQL without stream_context before 4? (which is _time:5m MarsSendServiceImpl in your case)?
A: The count of query is `86349`
![image](https://github.com/user-attachments/assets/9d308b0c-b45f-477c-8acb-ae6209dc66ae)

Q2: how many memory resource did you allocate to this VictoriaLogs instance?
A: The default configuration
![image](https://github.com/user-attachments/assets/faf1ef1e-9081-441f-b4c6-aa3c13d45909)

@jiekun

**Comment by jiekun** [comment]
Before addressing the issue, please allow me to provide an explanation of the steps involved in the `stream_context` pipe process:
1. retrieve all logs that fulfill the condition `_time:5m MarsSendServiceImpl` and group them by `stream` (`stream_id`).
2. gather **all** data for each stream (a.k.a: without time range filter). This step might be time-consuming if a stream contains a significant amount of data.
3. Go through the logs from step2. For each log from step 1, append 4 logs in the output based on their timestamps.

So to speed up the query execution (without changing the code), you can:
- reduce the number of streams generated in step 1 (by adjusting the condition before `stream_context`).
- reduce the data volume of each relevant stream (might be difficult for existing data).

It's definitely beneficial to optimize step 2 by fetching a smaller amount of data, such as by implementing a suitable time range filter. However, this is not a trivial fix and need discussion. (orz

**Comment by worker24h** [comment]
![image](https://github.com/user-attachments/assets/7455a4cf-0fbf-4076-bffb-3b374921f9ba)
OK，I try to special stream (e.g. {pod="xxx"})，unlucky， the response is still timeout

Is there a general reason for the problem？Sorry，I am reading the code and can't offer more useful info.

**Comment by worker24h** [comment]
> As an update, we haven't reached a conclusion on this issue in today's meeting. We are still making progress to find a better way to optimize it. Sorry for the delay.
> 
> Please let me know how it performs on some simple streams with a small data volume so that we can confirm the cause.

Yes, you are right. When I selected a small stream data,  it is so quickly.
![image](https://github.com/user-attachments/assets/5b8787e4-b3d6-41a5-acd3-69696fbca2e4)

I am debugging it on today, I found this line code which it has a problem.  
For example, the size of `contextRows` is `68022`,  the callback `writeBlock` will be executed by different goroutines, then the times of executed will be `68022 * N` (N ===>  the number of goroutines)

![image](https://github.com/user-attachments/assets/83520f05-bbee-45b1-b2c6-00d861d8bcf0)

**Comment by jiekun** [comment]
Actually most time are spent in this function. 

https://github.com/VictoriaMetrics/VictoriaMetrics/blob/66b2987f49dc5270e45f8a07694b3635c2eb7c81/lib/logstorage/pipe_stream_context.go#L157

```go
	writeBlock := func(_ uint, br *blockResult) {
		...
		for i := range contextRows {  // target rows
			...
			for j, timestamp := range timestamps { // all rows in a stream
				// update before & after in heap
				...
			}
		}
	}
```

For example, if you filtered out `N` logs from a stream containing `M` logs, VictoriaLogs will need to execute the logic inside this for-loop `M * N` times (`O(M*N)`).

Let's simplify the context of this problem:
- There are two **sorted** arrays, `A1` and `A2`, each with `N` and `M` items.
- For each item (`A1[i]`) in `A1`, find `x` items in `A2` that are not greater than `A1[i]` and closest to `A1[i]`.
- For each item (`A1[i]`) in A1, find `y` items in `A2` that are not less than `A1[i]` and closest to `A1[i]`.

I'm trying to use two pointers like this:
```go
                // `before` logic
                i, j := 0, 0
		for i < len(timestamps) && j < len(contextRows) {
			if timestamps[i] >= contextRows[j] {
                                // the previous 'x' items are the ones we are looking for.
				... // append previous 'x' items
				i++
			} else {
				j++
			}
		}

                // do the same logic for `after`, in reverse direction
```

The complexity for this method is `O((M+N)*(x+y))` ≈ `O(M+N)` when `x+y` is much smaller than `M+N`.

@valyala I'd like to have suggestions. Tried to optimize in this way but the result for `before` seems correct, but for `after`, it's not identical as what we currently have.

**Comment by worker24h** [comment]
> Actually most time are spent in this function. 
> 
> 
> 
> https://github.com/VictoriaMetrics/VictoriaMetrics/blob/66b2987f49dc5270e45f8a07694b3635c2eb7c81/lib/logstorage/pipe_stream_context.go#L157
> 
> 
> 
> ```go
> 
> 	writeBlock := func(_ uint, br *blockResult) {
> 
> 		...
> 
> 		for i := range contextRows {  // target rows
> 
> 			...
> 
> 			for j, timestamp := range timestamps { // all rows in a stream
> 
> 				// update before & after in heap
> 
> 				...
> 
> 			}
> 
> 		}
> 
> 	}
> 
> ```
> 
> 
> 
> For example, if you filtered out `N` logs from a stream containing `M` logs, VictoriaLogs will need to execute the logic inside this for-loop `M * N` times (`O(M*N)`).
> 
> 
> 
> Let's simplify the context of this problem:
> 
> - There are two **sorted** arrays, `A1` and `A2`, each with `N` and `M` items.
> 
> - For each item (`A1[i]`) in `A1`, find `x` items in `A2` that are **not greater than** `A1[i]` and closest to `A1[i]`.
> 
> - For each item (`A1[i]`) in `A1`, find `y` items in `A2` that are **not less than** `A1[i]` and closest to `A1[i]`.
> 
> 
> 
> I'm trying to use two pointers like this:
> 
> ```go
> 
>                 // `before` logic
> 
>                 i, j := 0, 0
> 
> 		for i < len(timestamps) && j < len(contextRows) {
> 
> 			if timestamps[i] >= contextRows[j] {
> 
>                                 // the previous 'x' items are the ones we are looking for.
> 
> 				... // append previous 'x' items
> 
> 				i++
> 
> 			} else {
> 
> 				j++
> 
> 			}
> 
> 		}
> 
> 
> 
>                 // do the same logic for `after`, in reverse direction
> 
> ```
> 
> 
> 
> The complexity for this method is `O((M+N)*(x+y))` ≈ `O(M+N)` when `x+y` is much smaller than `M+N`.
> 
> 
> 
> @valyala I'd like to have suggestions. Tried to optimize in this way. The result for `before` seems correct, but for `after`, it's not identical as what we currently have.

Hi, I think that the predicates `limit` and `time range` should be pushdown before the up logic which 
it is function `writeBlock`.

So, the scanned data will be much less. This approach is also commonly used for databases, e.g. MySQL

**Comment by valyala** [comment]
FYI, the commit bddb0e369f7b815046c3935d3aee6134bb35f889 improves performance of the [`stream_context` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#stream_context-pipe) when it is applied to logs, which belong to [log streams](https://docs.victoriametrics.com/victorialogs/keyconcepts/#stream-fields) with big number of logs (e.g. more than a million of logs). This commit will be included in the next release. In the mean time it is possible to build VictoriaLogs from source code according to [these instructions](https://docs.victoriametrics.com/victorialogs/quickstart/#building-from-source-code) and verifying whether this improves performance of the `stream_context` pipe.

**Comment by valyala** [comment]
> I had read the source code of Loki, the log context was showed in [now - 1h, now + 1h]. Your case typically doesn’t happen

This is a good idea! The commit 08af80ebe02cb0d299e7fc6e35cb6c48469fdcbd sets the time window for searching for surrounding logs in the [log stream](https://docs.victoriametrics.com/victorialogs/keyconcepts/#stream-fields) to one hour. This time window can be changed via `time_window <duration>` option in the `stream_context` pipe. For example, if you want to search for surrounding logs on a one year window, then the following `stream_context` pipe can be used:

```
_time:5m error | stream_context before 10 time_window 1y
```

This commit will be included in the next release of VictoriaLogs.

**Comment by worker24h** [comment]
@valyala @jiekun Congratulations, I will test it with the improvement, tomorrow.
Here, I have a problem, do you help me to resolve it? 

![image](https://github.com/user-attachments/assets/c0143c8a-c773-4670-af47-36225afa6797)

The variable `rows` is not order with timestamps? Why is it iterator by `for`?

**Comment by valyala** [comment]
> The variable rows is not order with timestamps? Why is it iterator by for?

The `rows` contains timestamps for the found surrounding log entries for the given log entry, which is found by the original query. This loops seeks for the minimum and the maximum timestamp across `rows`. The found time range is used as [time filter](https://docs.victoriametrics.com/victorialogs/logsql/#time-filter) in the query, which selects the needed fields for the surrounding logs on the found time range.

It is faster to iterate over `rows` for finding the minimum and the maximum timestamp than sorting `rows` and then selecting the first and the last item, since the complexity of the first case (`O(N)`) is smaller than the complexity of the second case (`O(N*log(N))`).

**Comment by valyala** [comment]
All the improvements for the [`stream_context` pipe](https://docs.victoriametrics.com/victorialogs/logsql/#stream_context-pipe) mentioned above are included in the release [v1.3.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/). Closing the issue as resolved.

@worker24h , please open a new issue and post a link to this issue there if you still see performance issues with the `stream_context` pipe in the `v1.3.0-victorialogs` release.
